internal/validators/os/os_validator.go

```diff
@@ -0,0 +1,126 @@
+// Package os handles OS image rule reconciliation.
+package os
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/canonical/gomaasclient/entity"
+	mapset "github.com/deckarep/golang-set/v2"
+	"github.com/go-logr/logr"
+
+	"github.com/validator-labs/validator-plugin-maas/api/v1alpha1"
+	"github.com/validator-labs/validator-plugin-maas/internal/constants"
+	vapi "github.com/validator-labs/validator/api/v1alpha1"
+	"github.com/validator-labs/validator/pkg/types"
+	"github.com/validator-labs/validator/pkg/util"
+)
+
+// ImageAPI is an interface for interacting with the Maas API
+type imageAPI interface {
+	ListOSImages() ([]entity.BootResource, error)
+}
+
+// ImageRulesService is a service for reconciling OS image rules
+type ImageRulesService struct {
+	log logr.Logger
+	api imageAPI
+}
+
+// NewImageRulesService returns a ImageRulesService
+func NewImageRulesService(log logr.Logger, api imageAPI) *ImageRulesService {
+	return &ImageRulesService{
+		log: log,
+		api: api,
+	}
+}
+
+// ReconcileMaasInstanceImageRules reconciles a MaaS instance image rule from the MaasValidator config
+func (s *ImageRulesService) ReconcileMaasInstanceImageRules(rule v1alpha1.OSImageRule) (*types.ValidationRuleResult, error) {
+
+	vr := buildValidationResult(rule)
+
+	brs, err := s.api.ListOSImages()
```

Comment:
suggestion: move the call to ListOSImages into findBootResources and make findBootResources a method so it can access `s.api`.